Add some basic HNSW graph checks to CheckIndex

benchaplin opened the pull request:
### Description

This is a first pass on #13724 - I've added the following checks:
- verify neighbor list is in order 
- verify no duplicate neighbors

The code was adapted from [KnnGraphTester](https://github.com/mikemccand/luceneutil/blob/3a6235c2038de6c6d5d0575d8f58cf06c2836793/src/main/knn/KnnGraphTester.java#L521) in luceneutil.

Here's some test results from an index containing the cohere-wikipedia-docs-768d dataset:
```
    test: open reader.........OK [took 0.010 sec]
    test: check integrity.....OK [took 2.207 sec]
    test: check live docs.....OK [took 0.000 sec]
    test: field infos.........OK [2 fields] [took 0.000 sec]
    test: field norms.........OK [0 fields] [took 0.000 sec]
    test: terms, freq, prox...    test: stored fields.......OK [1500000 total field count; avg 1.0 fields per doc] [took 0.398 sec]
    test: term vectors........OK [0 total term vector count; avg 0.0 term/freq vector fields per doc] [took 0.000 sec]
    test: docvalues...........OK [0 docvalues fields; 0 BINARY; 0 NUMERIC; 0 SORTED; 0 SORTED_NUMERIC; 0 SORTED_SET; 0 SKIPPING INDEX] [took 0.000 sec]
    test: points..............OK [0 fields, 0 points] [took 0.000 sec]
    test: vectors.............OK [1 fields, 1500000 vectors] [took 0.497 sec]
    test: hnsw graph..........OK [4 levels, 1547684 nodes (over all levels)] [took 0.485 sec]

No problems were detected with this index.
```

I manually corrupted some neighbors lists in order to check the failure cases:
```
    test: open reader.........OK [took 0.012 sec]
    test: check integrity.....OK [took 3.633 sec]
    test: check live docs.....OK [took 0.000 sec]
    test: field infos.........OK [2 fields] [took 0.000 sec]
    test: field norms.........OK [0 fields] [took 0.000 sec]
    test: terms, freq, prox...    test: stored fields.......OK [1500000 total field count; avg 1.0 fields per doc] [took 0.393 sec]
    test: term vectors........OK [0 total term vector count; avg 0.0 term/freq vector fields per doc] [took 0.000 sec]
    test: docvalues...........OK [0 docvalues fields; 0 BINARY; 0 NUMERIC; 0 SORTED; 0 SORTED_NUMERIC; 0 SORTED_SET; 0 SKIPPING INDEX] [took 0.000 sec]
    test: points..............OK [0 fields, 0 points] [took 0.000 sec]
    test: vectors.............OK [1 fields, 1500000 vectors] [took 0.504 sec]
    test: hnsw graph..........ERROR: org.apache.lucene.index.CheckIndex$CheckIndexException: Neighbors out of order for node 31142: 839445<1326517 1st=33190
org.apache.lucene.index.CheckIndex$CheckIndexException: Neighbors out of order for node 31142: 839445<1326517 1st=33190
	at org.apache.lucene.core@11.0.0-SNAPSHOT/org.apache.lucene.index.CheckIndex.testHnswGraph(CheckIndex.java:2798)
	at org.apache.lucene.core@11.0.0-SNAPSHOT/org.apache.lucene.index.CheckIndex.testSegment(CheckIndex.java:1109)
	at org.apache.lucene.core@11.0.0-SNAPSHOT/org.apache.lucene.index.CheckIndex.checkIndex(CheckIndex.java:806)
	at org.apache.lucene.core@11.0.0-SNAPSHOT/org.apache.lucene.index.CheckIndex.checkIndex(CheckIndex.java:576)
	at org.apache.lucene.core@11.0.0-SNAPSHOT/org.apache.lucene.index.CheckIndex.doCheck(CheckIndex.java:4479)
	at org.apache.lucene.core@11.0.0-SNAPSHOT/org.apache.lucene.index.CheckIndex.doMain(CheckIndex.java:4316)
	at org.apache.lucene.core@11.0.0-SNAPSHOT/org.apache.lucene.index.CheckIndex.main(CheckIndex.java:4248)
```

I intend to dig into the HNSW creation and think about more checks that would be useful here, but I wanted to start small with these neighbor checks only.

benchaplin committed:
Inspect Hnsw graph in CheckIndex

benchaplin committed:
Formalize integrity check for HnswGraph

benchaplin committed:
Clean up

benchaplin committed:
Abstract the HNSW codec

benchaplin committed:
Increment status during traversal

benchaplin committed:
Add condition to only run HNSW check when applicable

benchaplin committed:
Tidy

tteofili commented:
this might also check for graph connectivity, see #12627

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/CheckIndex.java:
```diff
@@ -406,6 +411,21 @@ public static final class VectorValuesStatus {
+      HnswGraphStatus() {}
+
+      /** Number of nodes in the graph */
+      public int hnswGraphSize;
```
Maybe rename to `graphNumNodes` or so?  `size` is a bit ambiguous/tricky, and since `hnsw` is in the class name we can prolly leave it off the member name?

benchaplin replied:
Good point - renamed it `totalNumNodes` to be extra clear.

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/CheckIndex.java:
```diff
@@ -406,6 +411,21 @@ public static final class VectorValuesStatus {
+      public int hnswGraphSize;
+
+      /** Number of levels of the graph */
+      public int hsnwGraphNumLevels;
```
`graphNumLevels`?

Naming is the hardest part!

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/CheckIndex.java:
```diff
@@ -2746,6 +2769,84 @@ public static Status.VectorValuesStatus testVectors(
+  private static HnswGraph getHnswGraph(CodecReader reader) throws IOException {
+    KnnVectorsReader vectorsReader = reader.getVectorReader();
+    if (vectorsReader instanceof PerFieldKnnVectorsFormat.FieldsReader) {
+      vectorsReader = ((PerFieldKnnVectorsFormat.FieldsReader) vectorsReader).getFieldReader("knn");
```
Hmm the field can be anything (not necessarily `knn` like `luceneutil` uses)?  Can we change this so we iterate `FieldInfos` and for any field that has KNN vectors enabled (`FieldInfo.hasVectorValues`, hmm but also has the HNSW graph (since I think a field can be only "flat" vectors?)) we do this check?

benchaplin replied:
Thanks, I didn't quite understand fields when I wrote this - I think I get it now. Alright, I've done what you suggested (as is also done in `testVectors`) and iterated over `FieldInfos`, performing the check only when it applies. 

Because we might now parse several HNSW graphs, I've restructured the status object to support per-graph data. Successful output will now look like:

```
    test: open reader.........OK [took 0.010 sec]
    test: check integrity.....OK [took 2.216 sec]
    test: check live docs.....OK [took 0.000 sec]
    test: field infos.........OK [2 fields] [took 0.000 sec]
    test: field norms.........OK [0 fields] [took 0.000 sec]
    test: terms, freq, prox...    test: stored fields.......OK [1500000 total field count; avg 1.0 fields per doc] [took 0.390 sec]
    test: term vectors........OK [0 total term vector count; avg 0.0 term/freq vector fields per doc] [took 0.000 sec]
    test: docvalues...........OK [0 docvalues fields; 0 BINARY; 0 NUMERIC; 0 SORTED; 0 SORTED_NUMERIC; 0 SORTED_SET; 0 SKIPPING INDEX] [took 0.000 sec]
    test: points..............OK [0 fields, 0 points] [took 0.000 sec]
    test: vectors.............OK [1 fields, 1500000 vectors] [took 0.496 sec]
    test: hnsw graphs.........OK [2 fields: (field name: knn1, levels: 4, total nodes: 1547684), (field name: knn2, levels: 4, total nodes: 1547684)] [took 0.979 sec]
```

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/CheckIndex.java:
```diff
@@ -2746,6 +2769,84 @@ public static Status.VectorValuesStatus testVectors(
+            }
+            status.hnswGraphSize++;
+          }
+          status.hsnwGraphNumLevels++;
```
Instead of incrementing here, you could just set it to `numLevels` from above?

benchaplin replied:
I could, but wanted to respect the object as a "status" - so if a check fails, `numLevels` will indicate on which level it failed. However we don't print the status if an exception is thrown. What do you think?

mikemccand reviewed (commented):
Thank you for tackling this @benchaplin!  This is a blind spot for `CheckIndex` today, which is scary.  I left some minor comments...

mikemccand commented:
> this might also check for graph connectivity, see #12627

+1 -- this is a tricky thing about these HNSW graphs (that they are not necessarily born connected but rather must be stitched up somehow afterwards).

But maybe do this separately?  I'd love to just get this initial coverage in first...

mikemccand commented:
Actually, `CheckIndex` does have some coverage for vectors and KNN graph (it confirms it can enumerate all vectors, and also runs some searches on it if it's not just flat vectors (`checkByte/FloatVectorValues`) , so it's not completely blind :)

benchaplin committed:
Rename status vars, iterate over fields performing HNSW check where it applies

benchaplin committed:
Handle status per HNSW graph

benchaplin committed:
Format message to handle empty results

msokolov commented on lucene/core/src/java/org/apache/lucene/index/CheckIndex.java:
```diff
@@ -2746,6 +2781,106 @@ public static Status.VectorValuesStatus testVectors(
+          hnswGraph.seek(level, node);
+          int nbr, lastNeighbor = -1, firstNeighbor = -1;
+          while ((nbr = hnswGraph.nextNeighbor()) != NO_MORE_DOCS) {
+            if (firstNeighbor == -1) {
```
perhaps at some point we could also check for uniqueness of the neighbors, and also check that the neighbors are in range [0, numGraphNodes], and finally on levels > 0, we would want to assert that the neighbors are on this level. But this can all be done separately; maybe we could add a comment here?

benchaplin replied:
Sounds good, I'll add a TODO (and add those checks in a followup PR). A couple questions first: 
- Doesn't (neighbors are in order) and (neighbors are not repeated) verify uniqueness?
- Also, why wouldn't we assert neighbors are on this level for level 0?

msokolov commented:
> Doesn't (neighbors are in order) and (neighbors are not repeated) verify uniqueness?

duh, yes :)
    
    > Also, why wouldn't we assert neighbors are on this level for level 0?

yes, we should, it's just that for level 0 this is the same as node in [0, size) wheras other levels are sparse - they have the same max = size-1, but not every node is included

benchaplin committed:
Add TODO for more checks

benchaplin committed:
Log HNSW report with per-level node count and connectivity

benchaplin committed:
Fix prints

benchaplin commented:
I've added @msokolov's node-in-range check + neighbor-on-same-level check, and @tteofili's connectivity check. I chose not to throw an exception for disconnectedness, as I've found there's often a couple disconnected nodes at level 0 (at least in my indices). Maybe we could throw if connectivity is below some threshold? 

So overall, exceptions are thrown if:
- A node is not in the range [0, graphSize - 1]
- A node's neighbor is not on the same level as the node
- A node's neighbors are out of order 
- A node's neighbors contain duplicates

I implemented a more detailed report in the logs: node counts and connectivity are printed for each level of the graph, e.g.
```
test: open reader.........OK [took 0.012 sec]
test: check integrity.....OK [took 0.489 sec]
test: check live docs.....OK [took 0.000 sec]
test: field infos.........OK [2 fields] [took 0.000 sec]
test: field norms.........OK [0 fields] [took 0.000 sec]
test: terms, freq, prox...    test: stored fields.......OK [200000 total field count; avg 1.0 fields per doc] [took 0.055 sec]
test: term vectors........OK [0 total term vector count; avg 0.0 term/freq vector fields per doc] [took 0.000 sec]
test: docvalues...........OK [0 docvalues fields; 0 BINARY; 0 NUMERIC; 0 SORTED; 0 SORTED_NUMERIC; 0 SORTED_SET; 0 SKIPPING INDEX] [took 0.000 sec]
test: points..............OK [0 fields, 0 points] [took 0.000 sec]
test: vectors.............OK [1 fields, 200000 vectors] [took 0.111 sec]
test: hnsw graphs.........OK [1 fields] [took 0.317 sec]
  hnsw field name: knn
    level 3: 11 nodes, 11/11 connected
    level 2: 141 nodes, 141/141 connected
    level 1: 6022 nodes, 6022/6022 connected
    level 0: 200000 nodes, 199999/200000 connected
```

msokolov commented on lucene/core/src/java/org/apache/lucene/index/CheckIndex.java:
```diff
@@ -2746,6 +2785,191 @@ public static Status.VectorValuesStatus testVectors(
+              status.hnswGraphsStatusByField.size(),
+              nsToSec(System.nanoTime() - startNS)));
+      printHnswInfo(infoStream, status.hnswGraphsStatusByField);
+    } catch (Throwable e) {
```
I think it should be enough to catch  Exception here and let an outer frame handle Errors

benwtrent replied:
agreed, only do exceptions here.

msokolov reviewed (commented):
This looks great - I just have one nit about the exception handling, otherwise I think it's ready to push

benwtrent commented on lucene/core/src/java/org/apache/lucene/index/CheckIndex.java:
```diff
@@ -2746,6 +2785,191 @@ public static Status.VectorValuesStatus testVectors(
+      if (fieldInfos.hasVectorValues()) {
+        for (FieldInfo fieldInfo : fieldInfos) {
+          if (fieldInfo.hasVectorValues()) {
+            if (vectorsReader instanceof PerFieldKnnVectorsFormat.FieldsReader) {
+              KnnVectorsReader fieldReader =
+                  ((PerFieldKnnVectorsFormat.FieldsReader) vectorsReader)
+                      .getFieldReader(fieldInfo.name);
+              if (fieldReader instanceof HnswGraphProvider) {
+                HnswGraph hnswGraph = ((HnswGraphProvider) fieldReader).getGraph(fieldInfo.name);
+                testHnswGraph(hnswGraph, fieldInfo.name, status);
+              }
```
```suggestion
            if (vectorsReader instanceof PerFieldKnnVectorsFormat.FieldsReader fieldsReader) {
              KnnVectorsReader fieldReader = fieldsReader.getFieldReader(fieldInfo.name);
              if (fieldReader instanceof HnswGraphProvider graphProvider) {
                HnswGraph hnswGraph = graphProvider.getGraph(fieldInfo.name);
                testHnswGraph(hnswGraph, fieldInfo.name, status);
              }
```

benchaplin replied:
thanks, much nicer!

benwtrent reviewed (commented):
One minor modernization comment.

mikemccand commented:
Thank you for persisting on this important change @benchaplin!

I applied this PR to my local Lucene clone and ran `CheckIndex` on the vector index created by [last night's nightly Lucene benchmarks](https://benchmarks.mikemccandless.com/) index but there was no new output covering the HNSW graph.

I think the reason is because the benchmarks [create an index using `Lucene99HnswVectorsFormat`](https://github.com/mikemccand/luceneutil/blob/main/src/main/perf/Indexer.java#L477-L493) without the `PerField` wrapper.

I think it'd be easy to also handle this case in your change?  In addition to checking if the `PerFieldKnnVectorsFormat` is in use, you could check whether the `reader.getVectorReader() instanceof HnswGraphProvider`?  So both routes (per-field case and not per-field case) could yield an `HnswGraphProvider` which you can then check for integrity ...

It's OK to do this in a follow-on issue too... either way.

benchaplin committed:
Handle instance where PerFieldKnnVectorsFormat is not used

msokolov commented:
Ooh, I was wondering about this `PerFieldKnnVectorsFormat` case - thanks for testing @mikemccand

mikemccand commented:
> Let me know if you're able to test this revision!

Yay, I just tested again and I see HNSW output, and the nightly benchy index passed all the new tests here!  Thanks @benchaplin.

```
    test: hnsw graphs.........OK [1 fields] [took 0.804 sec]
      hnsw field name: vector
        level 4: 2 nodes, 2/2 connected
        level 3: 85 nodes, 85/85 connected
        level 2: 1808 nodes, 1808/1808 connected
        level 1: 31568 nodes, 31565/31568 connected
        level 0: 497804 nodes, 497794/497804 connected
```

msokolov commented:
FYI @benchaplin this is merged now, but then I realized there is no CHANGES entry -- would you like to add one? After that, we can probably also backport to `branch_10x` using cherry-picks

benchaplin commented:
Thanks @mikemccand and @msokolov! And oh, forgot about the CHANGES - here's an entry (I listed under 10.2.0) https://github.com/apache/lucene/pull/14120.
